### PR TITLE
Support `returnGeneratedValues` for `INSERT` statement

### PR DESCRIPTION
**Pull request by mirromutth**
MySQL only support `SELECT LAST_INSERT_ID()` when last query has changed any values of `AUTO_INCREMENT`

### CODE REVIEW

**Comment by oshai** [comment]
@mp911de - can you please review that this is the expected api?

File: r2dbc-mysql/src/main/java/SimpleStatement.kt [hunk 1]
```diff
@@ -45,9 +47,40 @@ class SimpleStatement(private val clientSupplier: Supplier<JasyncConnection>, pr
         return this
     }
 
+    override fun returnGeneratedValues(vararg columns: String): Statement {
+        if (columns.any { !it.equals("LAST_INSERT_ID", true) }) {
```
Comment (mp911de) [inline r2dbc-mysql/src/main/java/SimpleStatement.kt]: The intention of `columns` is to specify a desired column name for generated keys. So the first column name should be used as an alias (`SELECT LAST_INSERT_ID() as $columns[0]`). See [SQL Server implementation](https://github.com/r2dbc/r2dbc-mssql/blob/master/src/main/java/io/r2dbc/mssql/GeneratedValues.java#L112) for further details.

File: r2dbc-mysql/src/main/java/SimpleStatement.kt [hunk 2]
```diff
@@ -45,9 +47,40 @@ class SimpleStatement(private val clientSupplier: Supplier<JasyncConnection>, pr
...
+
+        return this
+    }
+
     override fun execute(): Publisher<out Result> {
         val jasyncConnection = clientSupplier.get()
+        val queried = queryExecute(jasyncConnection)
+
+        if (selectLastInsertId) {
+            return queried.flatMap { result ->
```
Comment (mp911de) [inline r2dbc-mysql/src/main/java/SimpleStatement.kt]: Is it possible to combine `SELECT LAST_INSERT_ID()` with the insert statement? E.g. appending `SELECT LAST_INSERT_ID()` to the insert query so we get just a single server roundtrip?

File: r2dbc-mysql/src/main/java/SimpleStatement.kt [hunk 3]
```diff
@@ -62,7 +95,7 @@ class SimpleStatement(private val clientSupplier: Supplier<JasyncConnection>, pr
             } else {
                 jasyncConnection.sendQuery(this.sql)
             }
-            r.toMono().map { JaysncResult(it.rows, it.rowsAffected) }
+            r.toMono()
```
Comment (oshai) [inline r2dbc-mysql/src/main/java/SimpleStatement.kt]: I think the way to go (to save the additional roundtrip) is to cast the result here to `MySQLQueryResult` and use the `lastInsertId` property.

**Comment by mp911de** [comment]
I'll come back to this PR later today.

File: r2dbc-mysql/src/main/java/SimpleStatement.kt (hunk 2 above)
Comment (mp911de) [reply r2dbc-mysql/src/main/java/SimpleStatement.kt]: There are two checks for `selectLastInsertId`, a single check is sufficient. I also slightly changed the code:

```kotlin
         if (selectLastInsertId) {
            return queried.flatMap { result ->
                    jasyncConnection.sendQuery("SELECT LAST_INSERT_ID() AS $generatedKeyName")
                        .toMono()
                        .map { JaysncResult(it.rows, result.rowsAffected) }
            }
        }
```

File: r2dbc-mysql/src/main/java/SimpleStatement.kt (hunk 1 above)
Comment (mp911de) [reply r2dbc-mysql/src/main/java/SimpleStatement.kt]: That can be simplified to:

```kotlin
        if (columns.size == 1) {
            generatedKeyName = columns[0]
        }

        if (columns.size > 1) {
            throw IllegalArgumentException("MySQL only supports a single generated value")
        }

        selectLastInsertId = true

        return this
```

File: r2dbc-mysql/src/main/java/SimpleStatement.kt [hunk 4]
```diff
@@ -12,6 +13,7 @@ class SimpleStatement(private val clientSupplier: Supplier<JasyncConnection>, pr
 
     private var isPrepared = false
     private var params: MutableMap<Int, Any?> = mutableMapOf()
+    private var selectLastInsertId: Boolean = false
```
Comment (mp911de) [inline r2dbc-mysql/src/main/java/SimpleStatement.kt]: We should also add `private var generatedKeyName: String = "LAST_INSERT_ID"` here.

**Comment by oshai** [comment]
@mp911de what about using the value returned by the insert statement as I suggested?

**Comment by mp911de** [comment]
From a client perspective, it's desirable to have some control over the generated key name. Otherwise, a client has always to inspect metadata for the results. So the idea behind the columns is to give clients an ability to indicate the desired column name. Drivers can pick it up and apply the hint.

It will work also without a name hint, but it is nicer with a name hint.

**Comment by oshai** [comment]
Ok. Let me know when it is ready to be merged.

**Comment by mp911de** [comment]
Ping @mirromutth

**Comment by oshai** [comment]
@mp911de - I added the changes you suggested, let me know if it looks good and I will merge it.

File: r2dbc-mysql/src/main/java/SimpleStatement.kt (hunk 4 above)
Comment (oshai) [reply r2dbc-mysql/src/main/java/SimpleStatement.kt]: added below

**Comment by mp911de** [comment]
Looks good now!

**Comment by oshai** [comment]
Thanks!

**Pull request merged**